# Before you start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [504]:
#!pip install -U scipy --user

In [505]:
# import numpy and pandas

In [506]:
import numpy as np
import pandas as pd

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the San Francisco. We will start by loading the dataset and examining its contents. 

In [507]:
# Your code here:
df = pd.read_csv(r'C:\Users\duyadong\Ironhack\Week_4\Day_3\Afternoon\lab-Hypothesis-Testing\data\Salaries.csv')

C:\Users\duyadong\AppData\Local\Temp\ipykernel_12952\1740134979.py:2: DtypeWarning: Columns (3,4,5,6,12) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(r'C:\Users\duyadong\Ironhack\Week_4\Day_3\Afternoon\lab-Hypothesis-Testing\data\Salaries.csv')


Examine the `salaries` dataset using the `head` function below.

In [508]:
# Your code here:
df2 = df.copy() # create a copy of original data

In [509]:
df2.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay OvertimePay   OtherPay Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.0  400184.25      NaN  567595.43         567595.43   
1  155966.02   245131.88  137811.38      NaN  538909.28         538909.28   
2  212739.13   106088.18    16452.6      NaN  335279.91         335279.91   
3    77916.0    56120.71   198306.9      NaN  332343.61         332343.61   
4   134401.6      9737.0  182234.59      NaN  326373.19         326373.19   

   Year  Notes         Agency Status  
0  2011    NaN  San Francisco    NaN  
1  2011    NaN  San Francisco    NaN  
2  2011    NaN  San Francisco    NaN  
3  2011    NaN  San Francisco    NaN  
4  2011    NaN  San Francisco    NaN

We see from looking at the `head` function that there is quite a bit of missing data. Get the amount of missing data in every column

In [510]:
# Your code here:

In [511]:
df2.isna().sum()

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                605
OvertimePay              0
OtherPay                 0
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Notes               148654
Agency                   0
Status              110535
dtype: int64

Get the shape of the dataframe

In [512]:
# Your code here

In [513]:
df2.shape

(148654, 13)

Given output of the previous two cells, drop the corresponding column and compute again the amount of missing values.

In [514]:
# Your code here
df2 = df2.drop(['Notes'],axis=1)
df2.shape

(148654, 12)

In [515]:
df2.isna().sum()

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                605
OvertimePay              0
OtherPay                 0
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Agency                   0
Status              110535
dtype: int64

Check out what are the possible values of the column "Status".

In [516]:
# Your code here
df2['Status'].unique()

array([nan, 'PT', 'FT'], dtype=object)

In [517]:
df2['Status'].value_counts(dropna=False) #I checked again the number of each value in Status,including NA

NaN    110535
FT      22334
PT      15785
Name: Status, dtype: int64

Drop any row with missing values in the "Status" column and compute again the number of missing values.

In [518]:
# Your code here
df2= df2.dropna(subset = ['Status'])

In [519]:
df2.shape

(38119, 12)

In [520]:
df2.isna().sum()

Id                  0
EmployeeName        0
JobTitle            0
BasePay             0
OvertimePay         0
OtherPay            0
Benefits            0
TotalPay            0
TotalPayBenefits    0
Year                0
Agency              0
Status              0
dtype: int64

Check out the types of each column and see if they make sense.

In [521]:
# Your code here
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 38119 entries, 110531 to 148653
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Id                38119 non-null  int64  
 1   EmployeeName      38119 non-null  object 
 2   JobTitle          38119 non-null  object 
 3   BasePay           38119 non-null  object 
 4   OvertimePay       38119 non-null  object 
 5   OtherPay          38119 non-null  object 
 6   Benefits          38119 non-null  object 
 7   TotalPay          38119 non-null  float64
 8   TotalPayBenefits  38119 non-null  float64
 9   Year              38119 non-null  int64  
 10  Agency            38119 non-null  object 
 11  Status            38119 non-null  object 
dtypes: float64(2), int64(2), object(8)
memory usage: 3.8+ MB


In [522]:
df2.head()

Id          EmployeeName                  JobTitle    BasePay  \
110531  110532           David Shinn            Deputy Chief 3  129150.01   
110532  110533            Amy P Hart         Asst Med Examiner  318835.49   
110533  110534  William J Coaker Jr.  Chief Investment Officer   257340.0   
110534  110535        Gregory P Suhr           Chief of Police  307450.04   
110535  110536  Joanne M Hayes-White    Chief, Fire Department   302068.0   

       OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
110531         0.0  342802.63  38780.04  471952.64         510732.68  2014   
110532    10712.95   60563.54  89540.23  390111.98         479652.21  2014   
110533         0.0    82313.7  96570.66  339653.70         436224.36  2014   
110534         0.0   19266.72  91302.46  326716.76         418019.22  2014   
110535         0.0   24165.44  91201.66  326233.44         417435.10  2014   

               Agency Status  
110531  San Francisco     PT  
110532  San Francisco     FT  
110533  San Francisco     PT  
110534  San Francisco     FT  
110535  San Francisco     FT

Do any type conversions and reset the index.

In [523]:
# Your code here

In [524]:
convert_dict = {'BasePay': float,
                'OvertimePay': float,
                'OtherPay':float,
                'Benefits':float
                }
df2 = df2.astype(convert_dict)
#convert the numbers into float

In [525]:
df2.info() #now they are all float

<class 'pandas.core.frame.DataFrame'>
Int64Index: 38119 entries, 110531 to 148653
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Id                38119 non-null  int64  
 1   EmployeeName      38119 non-null  object 
 2   JobTitle          38119 non-null  object 
 3   BasePay           38119 non-null  float64
 4   OvertimePay       38119 non-null  float64
 5   OtherPay          38119 non-null  float64
 6   Benefits          38119 non-null  float64
 7   TotalPay          38119 non-null  float64
 8   TotalPayBenefits  38119 non-null  float64
 9   Year              38119 non-null  int64  
 10  Agency            38119 non-null  object 
 11  Status            38119 non-null  object 
dtypes: float64(6), int64(2), object(4)
memory usage: 3.8+ MB


In [526]:
df2=df2.reset_index(drop=True)

In [527]:
df2.head()

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  
0  San Francisco     PT  
1  San Francisco     FT  
2  San Francisco     PT  
3  San Francisco     FT  
4  San Francisco     FT

Check out if "TotalPayBenefits" = "BasePay" + "OvertimePay" + "OtherPay" + "Benefits"

In [528]:
# Your code here
equality=(df2['TotalPayBenefits'] == df2['BasePay']+df2[ "OvertimePay"]+df2['OtherPay']+df2['Benefits'])

In [529]:
print(equality)

0         True
1        False
2         True
3        False
4         True
         ...  
38114     True
38115     True
38116     True
38117     True
38118     True
Length: 38119, dtype: bool


What is the percetage of employees for which the previous assumption is not True?

In [530]:
# Your code here:
equality.value_counts(normalize=True)*100

True     73.708649
False    26.291351
dtype: float64

There are different departments in the city. List all departments and the count of employees in each department.

In [531]:
# Your code here:
df2.groupby(by='JobTitle')['Id'].count().reset_index(name='Empolyees')

JobTitle  Empolyees
0       ACPO,JuvP, Juv Prob (SFERS)          1
1      ASR Senior Office Specialist         22
2              ASR-Office Assistant         15
3                     Account Clerk         93
4                      Accountant I          2
..                              ...        ...
991      Wire Rope Cable Maint Sprv          1
992      Worker's Comp Supervisor 1          6
993  Worker's Compensation Adjuster         26
994           X-Ray Laboratory Aide         35
995              Youth Comm Advisor          1

[996 rows x 2 columns]

# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of **all FT workers is significantly different from $75/hr**. Get first the hourly wage by dividing "TotalPayBenefits" by 50 weeks (assuming 10 labour days of holidays) and by 40hrs (assuming a 40hrs week).

$$Hourly Wage = \frac{TotalPayBenefits}{1 year}\frac{1 year}{50 Week}\frac{1 Week}{40 hr}$$

Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [532]:
# Your code here: (compute the "Hourly_Wage")
df2['Hourly_wage']=df2['TotalPayBenefits']/(50*40)

In [533]:
df2.head()

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  Hourly_wage  
0  San Francisco     PT   255.366340  
1  San Francisco     FT   239.826105  
2  San Francisco     PT   218.112180  
3  San Francisco     FT   209.009610  
4  San Francisco     FT   208.717550

In [534]:
# Your code here: (Compute the mean hourly wage for all the "FT" employees)
ftwage=df2.loc[df2['Status']=='FT','Hourly_wage']
mean=ftwage.mean()
mean

69.264206407271

In [535]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider "FT" employees
from scipy.stats import sem
std_error = sem(ftwage)
t = (mean - 75) / std_error
t

-35.80631941460793

In [536]:
# Method 1: Critical value. Get the critical value and compare it against your statisttic.
# Your code here: 
df2['Status'].value_counts(dropna=False)

FT    22334
PT    15785
Name: Status, dtype: int64

In [537]:
from scipy.stats import t
t.ppf(0.025,22333)

-1.9600702128845597

In [538]:
t.ppf((1-0.025),22333)

1.9600702128845593

With the statistic method the t-statistics is smaller than the lower critical value,we should reject H0,so that all FT workers is significantly different from $75/hr.

In [539]:
# Method 2: Use the p-value method.
# Your code here:
t.cdf(-35.80631941460793,22333)

2.2589156167156547e-273

With the P value method,the P value is much more smaller than 0.025,so we reject the H0.

In [540]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:
from scipy.stats import ttest_1samp
ttest_1samp(ftwage,75)

Ttest_1sampResult(statistic=-35.80631941460526, pvalue=4.517831233839448e-273)

With the ttest_1samp function the p value is doubled because it's two-sided P value,and it's still much smaller than 0.025,so we reject the H0.

Are all the methods in agreement?

Yes,we reject H0 in all three methods.

We are also curious about salaries in the police force. The chief of police in Chicago claimed in a press briefing that salaries this year are **higher than last year's mean of $86000/year for all salaried employees** (use the column "TotalPayBenefits". Test  hypothesis using a 95% confidence interval.

Hint: Use apply and a lambda function to check in "Police" is in the "JobTitle" to get all the "Police" jobs.

In [541]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider "Police" employees
police= df2[df2['JobTitle'].str.contains('Police')]
salary_poli=police['TotalPayBenefits']
mean_poli =salary_poli.mean()
std_error_poli = sem(salary_poli)
t_poli = (mean_poli - 86000) / std_error_poli
t_poli


50.2529847421099

In [542]:
# Method 1: Critical value. Get the critical value and compare it against your statisttic.
# Your code here: 

In [543]:
mean_poli

148349.04212519721

In [544]:
salary_poli.shape

(1901,)

In [545]:
t.ppf((1-0.05),1900) #df = n-1

1.6456560042647304

t_poli(50.2529847421099) is bigger than critical value(1.6456560042647304) ->we reject the H0,so the salary is higher than last year.

In [546]:
# Method 2: Use the p-value method.
# Your code here:

In [547]:
1-t.cdf(50.2529847421099,1900)

0.0

With P value method,the p value(0.0) is lower than 0.05,so we reject H0.

In [548]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:
ttest_1samp(salary_poli,86000)

Ttest_1sampResult(statistic=50.252984742109945, pvalue=0.0)

With the third method,the P value is also 0.0,we reject the H0.

The workers from the "JobTitle" with the most employees have complained that their hourly wage is **less than $35/hour**. Using a one sample t-test, test this one-sided hypothesis at the 95% confidence level.

In [549]:
# Your code here: (Get the department which has most employees)
#df2['JobTitle'].mode()
df2['JobTitle'].value_counts().idxmax()

'Transit Operator'

In [550]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider the right "JobTitle" employees
trans= df2[df2['JobTitle']=='Transit Operator']
wage_trans=trans['Hourly_wage']
mean_trans =wage_trans.mean()
std_error_trans = sem(wage_trans)
t_trans = (mean_trans - 35) / std_error_trans
t_trans

19.064990324906578

In [551]:
# Method 1: Critical value. Get the critical value and compare it against your statisttic.
# Your code here: 
wage_trans.shape

(2479,)

In [552]:
t.ppf(0.05,2478)

-1.6454687772470247

the t_statistics(19.064990324906578) is higher than the critical value(-1.6454687772470247),we accept the H0,so the hourly wage is actually more than $35/h

In [553]:
# Method 2: Use the p-value method.
# Your code here:
t.cdf(19.064990324906578,2478)

1.0

the P value(1.0) is higher than 0.05,so we accept H0.

In [554]:
mean_trans

43.35949773900775

In [555]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:
ttest_1samp(wage_trans,35,alternative='less')
#the the alternative hypothesis is the mean is less than the population mean,so we add alternative = 'less' here to make it a one side problem.

Ttest_1sampResult(statistic=19.064990324906386, pvalue=1.0)

the P value(1.0) is higher than 0.05,so we accept H0.

# Challenge 3: To practice - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level.

In [556]:
# Method 1: Get the critical values which correspond to a 95% confidence.
# Your code here:
wage=df2['Hourly_wage']

In [557]:
from scipy.stats import norm
norm.interval(alpha=0.95,loc=np.mean(wage), scale=sem(wage)) 

C:\Users\duyadong\AppData\Local\Temp\ipykernel_12952\2625831091.py:2: DeprecationWarning: Use of keyword argument `alpha` for method `interval` is deprecated. Use first positional argument or keyword argument `confidence` instead.
  norm.interval(alpha=0.95,loc=np.mean(wage), scale=sem(wage))


(49.79826596983879, 50.463172698540085)

In [558]:
import statsmodels.stats.api as sms
sms.DescrStatsW(wage).tconfint_mean(alpha=0.05)#just another way doing it

(49.79825541308488, 50.463183255295725)

Now compute a 95% confidence interval for the hourly salary of all the Police employees.

In [559]:
# Your code here:
hour_poli= salary_poli/(50*40)
norm.interval(alpha=0.95,loc=np.mean(hour_poli), scale=sem(hour_poli)) 

C:\Users\duyadong\AppData\Local\Temp\ipykernel_12952\2645842802.py:3: DeprecationWarning: Use of keyword argument `alpha` for method `interval` is deprecated. Use first positional argument or keyword argument `confidence` instead.
  norm.interval(alpha=0.95,loc=np.mean(hour_poli), scale=sem(hour_poli))


(72.9586542074945, 75.3903879177025)

# Chi2 test

Now we want to know if the amount of full time "FT" and part time "PT" employees is equal between Lawers, Meds, Police, Firemen and other departments. 

Considering all the options in this groups of employees will be very time consuming. To simplify this process, create first a function that returns:

* "Policemen" if "Police" is found on "JobTitle"
* "Firemen" if "Fire" is found on "JobTitle"
* "Medical" if "Med" or "Nurse" is found on "JobTitle"
* "Lawyer" if "Attorney" is found on "JobTitle"
* "Other" in any other cases

Then, create a new column named "employee_group" that determines to which group belong the employee. 

In [640]:
# create a list of our conditions
# conditions = [
#     df2['JobTitle'].str.contains('Police'),
#     df2['JobTitle'].str.contains('Fire'),
#     df2['JobTitle'].str.contains('Med','Nurse'),
#     df2['JobTitle'].str.contains('Attorney'),
#     ]

# values = ['Policeman','Fireman','Medical','Lawyer']
# df2['employee_group'] = np.select(conditions, values,'Other')
# df2.head()

In [636]:
def group (x):
    if ('Police' in x):
        return 'Policeman'
    elif ('Fire' in x):
        return 'Fireman'
    elif ('Med' in x):
        return 'Medical'
    elif ('Nurse' in x):
        return 'Medical'
    elif ('Attorney' in x):
        return 'Lawyer'
    else:
        return 'Other'

In [637]:
df2['employee_group']=df2['JobTitle'].apply(group)

In [638]:
df2.head()

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  Hourly_wage employee_group  
0  San Francisco     PT   255.366340          Other  
1  San Francisco     FT   239.826105        Medical  
2  San Francisco     PT   218.112180          Other  
3  San Francisco     FT   209.009610      Policeman  
4  San Francisco     FT   208.717550        Fireman

Determine how many "PT" and "FT" employess have all the employees groups.

In [639]:
# Your code here: (Store the output dataframe into a new variable)
#df2.groupby(['employee_group','Status']).size()
#df2.value_counts(subset=['employee_group', 'Status']) 
#df2.groupby(['employee_group'])['Status'].count()
#df2.groupby(['employee_group','Status']).agg({'Status':'count'})
#crosstab=pd.crosstab(df2['employee_group'],df2['Status'],values=df2['Status'],aggfunc='count')
crosstab = pd.crosstab(df2['employee_group'], df2['Status']) 
crosstab

Status             FT     PT
employee_group              
Fireman          1333    178
Lawyer            317    102
Medical          1028   2889
Other           18126  12245
Policeman        1530    371

Now try compute the expected frequencies doing the calculations with the individual probabilities. Remember that the Chi2 test assumes that both variables (employee_group and FT/PT) are not related (therefore they are independent). Therefore, to compute the expected frequencies you need to compute the probability of each cell and multiply it by the number of observations. ie:

$$\nu(x,y) = p(x,y) * N = p(x) * p(y) * N$$

bear in mind that in general: $p(x,y)\neq p(x)*p(y)$; the equality will only be true if x and y are independent. However, the null hypotheses says that **x and y are independent.** but that's what we're assuming with the null hypotheses.

where "x" is the "employee_group" and "y" the (FT/PT). 

In [619]:
# Create an empty dataframe named "frequencies" to store the data.
# Your code here:
from scipy.stats import chi2_contingency
chi2_contingency(crosstab)

(1077.7974464881881,
 2.387092311770826e-233,
 3,
 array([[  885.29798788,   625.70201212],
        [  245.49295627,   173.50704373],
        [20089.40927097, 14198.59072903],
        [ 1113.79978488,   787.20021512]]))

In [620]:
frequency = pd.DataFrame()

In [621]:
df2.shape

(38119, 14)

In [605]:
ft=1333+317+333+18821+1530
ft

22334

In [603]:
# Your code here: Compute Expected frequency of being "Firemen" and "FT". Store the solution in a variable named "firemen_ft"
firemen_ft = ((1333+178)/38119)*(22334/38119)*38119
firemen_ft

885.2979878800596

In [606]:
# Your code here: Compute Expected frequency of being "Firemen" and "PT". Store the solution in a variable named "firemen_pt"
firemen_pt = ((1333+178)/38119)*((38119-22334)/38119)*38119
firemen_pt

625.7020121199403

In [607]:
# Your code here: Compute Expected frequency of being "Lawyers" and "FT". Store the solution in a variable named "lawyers_ft"
lawyers_ft = ((317+102)/38119)*(22334/38119)*38119
lawyers_ft

245.49295626852748

In [608]:
# Your code here: Compute Expected frequency of being "Lawyers" and "PT". Store the solution in a variable named "lawyers_pt"
lawyers_pt = ((317+102)/38119)*((38119-22334)/38119)*38119
lawyers_pt

173.5070437314725

In [641]:
# Your code here: Compute Expected frequency of being "Medical" and "FT". Store the solution in a variable named "medical_ft"
medical_ft = ((1028+2889)/38119)*(22334/38119)*38119
medical_ft

2294.9783047823917

In [642]:
# Your code here: Compute Expected frequency of being "Medical" and "PT". Store the solution in a variable named "medical_pt"
medical_pt = ((1028+2889)/38119)*((38119-22334)/38119)*38119
medical_pt

1622.021695217608

In [644]:
# Your code here: Compute Expected frequency of being "Other" and "FT". Store the solution in a variable named "other_ft"
other_ft = ((18126+12245)/38119)*(22334/38119)*38119
other_ft

17794.430966184842

In [645]:
# Your code here: Compute Expected frequency of being "Other" and "PT". Store the solution in a variable named "other_pt"
other_pt = ((18126+12245)/38119)*((38119-22334)/38119)*38119
other_pt

12576.569033815158

In [629]:
# Your code here: Compute Expected frequency of being "Policement" and "FT". Store the solution in a variable named "policemen_ft"
policemen_ft = ((1530+371)/38119)*(22334/38119)*38119
policemen_ft

1113.7997848841785

In [648]:
# Your code here: Compute Expected frequency of being "Policement" and "PT". Store the solution in a variable named "policemen_pt"
police_pt = ((1530+371)/38119)*((38119-22334)/38119)*38119
police_pt

787.2002151158215

* Store all the expected frequencies of "FT" employees in a list 
* Store all the "PT" employees into another list
* Create a dictionary with "FT" and "PT" as keys and as the values use the previous lists
* Create a dataframe with this dictionary using pd.DataFrame()

In [650]:
# Your code here:
ft=[885.2979878800596,245.49295626852748,2294.9783047823917,17794.430966184842,1113.7997848841785]
pt=[625.7020121199403,173.5070437314725,1622.021695217608,12576.569033815158,787.2002151158215]

In [652]:
dic={'FT':ft,'PT':pt}
dic=pd.DataFrame(dic)
dic

FT            PT
0    885.297988    625.702012
1    245.492956    173.507044
2   2294.978305   1622.021695
3  17794.430966  12576.569034
4   1113.799785    787.200215

Now use the "st.chi2_contingency()" from scipy.stats [documentation here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html) to conduct a Chi2 test to determine if the diferences between employee groups are statistically significant using a 95% confidence level. Hint: fill the function with a dataframe of actual frequencies.

In [653]:
# Your code here: (use the st.chi2_contingency() function from scipy.stats to compute:
# The Chi2 value
# The p-valueYea we 
# The expected frequencies.
# Ho: there is no relationship
# Ha: there is relationship differences
# p_value = P(table | Ho) = P(table | no relationship) = 1.51e-6 < 0.05
from scipy.stats import chi2_contingency
chi2_contingency(crosstab)

(2676.642333711905,
 0.0,
 4,
 array([[  885.29798788,   625.70201212],
        [  245.49295627,   173.50704373],
        [ 2294.97830478,  1622.02169522],
        [17794.43096618, 12576.56903382],
        [ 1113.79978488,   787.20021512]]))

Check if your expected frequencies aggree with the ones obtained with the st.chi2_contingency() function

Yes,they are exactly the same.